sdks/java/core/src/main/java/org/apache/beam/sdk/util/FlexibleBackoff.java

```diff
@@ -0,0 +1,202 @@
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.util;
+
+import static com.google.common.base.Preconditions.checkArgument;
+
+import com.google.api.client.util.BackOff;
+import org.joda.time.Duration;
+
+
+/**
+ * A {@link BackOff} for which the retry algorithm is customizable. By default the
+ * {@link FlexibleBackoff} uses exponential backoff (base exponent 1.5) with an initial backoff of
+ * 1 second. These parameters can be overridden with {@link FlexibleBackoff#withExponent(double)}
+ * and {@link FlexibleBackoff#withInitialBackoff(Duration)}, respectively, and the maximum backoff
+ * after exponential increase can be capped using {@link FlexibleBackoff#withMaxBackoff(Duration)}.
+ *
+ * <p>By default, and after modifying any of the above parameters, {@link FlexibleBackoff} does not
+ * limit the number of retries. To limit the backoff, the maximum total number of retries can be set
+ * using {@link FlexibleBackoff#withMaxRetries(int)}. The total time spent in backoff can be
+ * time-bounded as well by configuring {@link FlexibleBackoff#withMaxCumulativeBackoff(Duration)}.
+ * If either of these limits are reached, the {@link FlexibleBackoff#nextBackOffMillis()} will
+ * return {@link BackOff#STOP} to signal that no more retries should continue.
+ */
+public final class FlexibleBackoff implements BackOff {
+  private static final double DEFAULT_EXPONENT = 1.5;
+  private static final double DEFAULT_RANDOMIZATION_FACTOR = 0.5;
+  private static final Duration DEFAULT_MIN_BACKOFF = Duration.standardSeconds(1);
+  private static final Duration DEFAULT_MAX_BACKOFF = Duration.standardDays(1000);
+  private static final int DEFAULT_MAX_RETRIES = Integer.MAX_VALUE;
+  private static final Duration DEFAULT_MAX_CUM_BACKOFF = Duration.standardDays(1000);
+  // Customization of this backoff.
+  private final double exponent;
+  private final Duration initialBackoff;
+  private final Duration maxBackoff;
+  private final Duration maxCumulativeBackoff;
+  private final int maxRetries;
+  // Current state
+  private Duration currentCumulativeBackoff;
+  private int currentRetry;
+
+  /**
+   * Instantiates a {@link FlexibleBackoff} with the default exponential backoff and initial
+   * backoff.
+   *
+   * @see FlexibleBackoff
+   */
+  public static FlexibleBackoff of() {
```

Comment:
I think this would the right place to document defaults, rather than the class Javadoc.